File: python/src/thunder_book/ch06/mcts.py
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Generic, Optional, TypeVar

import fire
import numpy as np
from pydantic import BaseModel

from thunder_book.ch06.game import many_games
from thunder_book.ch06.maze_state import ActionFunc
from thunder_book.ch06.maze_state import SimulataneousMazeState as State
from thunder_book.ch06.monte_carlo import make_monte_carlo_f
from thunder_book.ch06.random_action import random_action

"""
def playout(state: State):
    if state.is_done():
        # 結局treeの作り方上、player_id 0でしかスコア計算しない
        return state.score(0)

    state.advance(
        random_action(state, 0),
        random_action(state, 1),
    )
    return playout(state)
"""


class MCTSParams(BaseModel):
    c: float
    expand_threshold: int


class Playout:
    def __init__(self, state: State) -> None:
        """上の playout でstate を渡す際にcopyしなかったバグを
        踏んだので, 絶対に copy するための工夫。
        """
        self.state = state.copy()

    def playout(self) -> float:
        if self.state.is_done():
            return self.state.score(0)

        self.state.advance(
            random_action(self.state, 0),
            random_action(self.state, 1),
        )
        return self.playout()


# Define a type variable that can be any subclass of BaseNode
T = TypeVar("T", bound="BaseNode")


class BaseNode(ABC, Generic[T]):
    def __init__(
        self,
        state: State,
        params: MCTSParams,
        root: Optional[BaseNode] = None,
        is_root: bool = False,
    ) -> None:
        self.params = params
        self.state = state.copy()
        self.w = 0
        self.n = 0
        self.child_nodes: list[T] = []
        self.root = root if root is not None else self
        assert root is not None or is_root, "root node must be set on non-root nodes"

    def _increment(self, value: float) -> None:
        self.w += value
        self.n += 1

    def __repr__(self) -> str:
        return f"{self.__class__.__name__}(w={self.w}, n={self.n})"

    @property
    def find(self) -> None:
        # calls where am i with self
        return print(self.root._where_am_i(self))

    def __str__(self) -> str:
        return self.root._where_am_i(self)

    @property
    def f(self) -> None:
        # helper for debugger
        return self.find

    def _where_am_i(self, target: BaseNode, depth: int = 0) -> str:
        # prints the staus from the root node
        # but adds a <<< to the current node
        stat = "__ " * depth
        mark = " <<<" if self == target else ""
        stat += self.__repr__() + mark + "\n"
        stat += "".join(c._where_am_i(target, depth + 1) for c in self.child_nodes)
        return stat

    def ucb1(self, t: float) -> float:
        # 子のノードの計算をするため、1-w が必要?
        return 1 - self.w / self.n + self.params.c * np.sqrt(2 * np.log(t) / self.n)

    def next_child_node(self) -> T:
        for child_node in self.child_nodes:
            if child_node.n == 0:
                return child_node

        # select best based on UCB1
        t = sum(c.n for c in self.child_nodes)
        ucb1s = [c.ucb1(t) for c in self.child_nodes]
        return self.child_nodes[np.argmax(ucb1s)]

    @abstractmethod
    def explore(self) -> float:
        NotImplemented()

    @abstractmethod
    def expand(self) -> None:
        NotImplemented()


# Node, OddNode, EvenNode で分ける方が
# 抽象化としてはキレイそう
# やったけど player0 じゃないとプレーできなくしてしまったので
# 自分自身とのプレーができない
# それもやりたいなら A_node, B_node とかにして
# 中身は player_id == 0/1 で分岐するようにするとかありそう


class EvenNode(BaseNode["OddNode"]):
    def __init__(
        self,
        state: State,
        params: MCTSParams,
        root: Optional[BaseNode] = None,
        is_root: bool = False,
    ) -> None:
        super().__init__(state, params, root, is_root=is_root)

    def expand(self) -> None:
        # call after player 0 explore finish
        legal_actions = self.state.legal_actions(0)
        self.child_nodes.clear()
        for action in legal_actions:
            self.child_nodes.append(OddNode(self.state, self.params, self.root, action))

    def explore(self) -> float:
        # evenは末端ノードではないので,必ずexpandする
        if not self.child_nodes:
            self.expand()

        value = self.next_child_node().explore()
        self._increment(value)
        return value


class OddNode(BaseNode["EvenNode"]):
    def __init__(self, state: State, params: MCTSParams, root: BaseNode, action0: int) -> None:
        super().__init__(state, params, root)
        # the action of player 0, just before this node
        self.action0 = action0

    def expand(self) -> None:
        legal_actions_opp = self.state.legal_actions(1)
        self.child_nodes.clear()
        for action1 in legal_actions_opp:
            self.child_nodes.append(EvenNode(self.state, self.params, self.root))
            assert isinstance(self.action0, int)
            self.child_nodes[-1].state.advance(self.action0, action1)

    def explore(self) -> float:
        # 奇数番目の場合,値を返す
        # 相手のスコアなので内部で　1-value を increment している
        if self.state.is_done():
            value = self.state.score(0)
            self._increment(1 - value)
            return value

        if self.child_nodes:
            value = self.next_child_node().explore()
            self._increment(1 - value)
            return value

        # no childs, return playout value
        value = Playout(self.state).playout()
        self._increment(1 - value)
        if self.n >= self.params.expand_threshold:
            self.expand()
        return value


def mcts_action(
    state: State,
    player_id: int,
    playout_number: int,
) -> int:
    # safeguard
    if player_id != 0:
        raise RuntimeError("player_id must be 0 for mcts")

    node = EvenNode(state, MCTSParams(c=1.0, expand_threshold=10), is_root=True)
    node.expand()
    for _ in range(playout_number):
        # breakpoint()
        node.explore()

    legal_actions = state.legal_actions(0)

    # get argmax of n
    num_searched = [c.n for c in node.child_nodes]
    return legal_actions[np.argmax(num_searched)]


def make_mcts_f(playout_number: int) -> ActionFunc:
    def mcts_f(state: State, player_id: int) -> int:
        return mcts_action(state, player_id, playout_number)

    return mcts_f


def mcts_vs_monte_carlo(num_playout=100, num_games=100):
    print(f"mcts vs monte carlo {num_playout}")

    monte_carlo_f = make_monte_carlo_f(num_playout)
    mcts_f = make_mcts_f(num_playout)

    actions_bw = (mcts_f, monte_carlo_f)
    win_rate = many_games(num_games, actions_bw, player_id=0, print_every=10)
    print(f"{win_rate=:.2f} for mcts vs monte carlo {num_playout}")


def mcts_vs_random_action(num_playout=100, num_games=100):
    print(f"mcts {num_playout} vs random")
    mcts_f = make_mcts_f(num_playout)
    actions_bw = (mcts_f, random_action)
    win_rate = many_games(num_games, actions_bw, player_id=0, print_every=10)
    print(f"{win_rate=:.2f} for mcts {num_playout} vs random")


def main(game="monte_carlo", *args, **kwargs):
    if game == "random":
        return mcts_vs_random_action(*args, **kwargs)
    if game == "monte_carlo":
        return mcts_vs_monte_carlo(*args, **kwargs)
    raise ValueError(f"{game=} is not supported")


if __name__ == "__main__":
    fire.Fire(main)

File: cpp/src/ch08/mcts.cc
#include <iostream>
#include <assert.h>
#include "mcts.h"
#include "random_action.h"
#include "time_keeper.h"

double Playout::playout()
{
    if (state_.is_done())
        return state_.teban_score();

    state_.advance(random_action(state_));
    return playout();
}

int mcts_action(const ConnectFourState &state,
                const int playout_number,
                const bool should_print)
{
    Node node = Node(state);
    node.expand();
    for (int i = 0; i < playout_number; i++)
    {
        node.evaluate();
        if (should_print)
            node.print_tree();
    }

    return node.best_action();
}

int mcts_action_timebound(const ConnectFourState &state,
                          const int64_t time_threshold)
{
    Node node = Node(state);
    node.expand();
    auto time_keeper = TimeKeeper(time_threshold);
    while (!time_keeper.is_time_over())
        node.evaluate();

    return node.best_action();
}

int Node::best_action()
{
    auto legal_actions = state_.legal_actions();
    int best_action_searched_number = -1;
    int best_action_index = -1;
    assert(legal_actions.size() == child_nodes_.size());

    // 試行回数の多いノードを選ぶ（いいノードは試行回数も多いから)
    for (int i = 0; i < (int)legal_actions.size(); i++)
    {
        int n = child_nodes_[i].n_;
        if (n > best_action_searched_number)
        {
            best_action_index = i;
            best_action_searched_number = n;
        }
    }
    return legal_actions[best_action_index];
}

void Node::_increment(double value)
{
    w_ += value;
    n_++;
}

double Node::evaluate()
{
    if (state_.is_done())
    {
        double value = state_.teban_score();
        _increment(value);
        return value;
    }

    if (child_nodes_.empty())
    {
        double value = Playout(state_).playout();
        _increment(value);
        if (n_ == EXPAND_THRESHOLD)
            expand();
        return value;
    }

    // has child
    double value = 1. - next_child_node().evaluate();
    _increment(value);
    return value;
}

void Node::expand()
{
    auto legal_actions = state_.legal_actions();
    child_nodes_.clear();
    for (const auto action : legal_actions)
    {
        child_nodes_.emplace_back(state_);
        child_nodes_.back().state_.advance(action);
    }
}

double Node::t_() const
{
    double t = 0;
    for (const auto &child_node : child_nodes_)
        t += child_node.n_;
    return t;
}

double Node::ucb1(double t) const
{
    using std::log;
    using std::sqrt;
    return 1 - w_ / n_ + (double)C * sqrt(2. * log(t) / n_);
}

Node &Node::next_child_node()
{
    for (auto &child_node : child_nodes_)
        if (child_node.n_ == 0)
            return child_node;

    // select best ucb1
    double t = t_();
    double best_value = -INF;
    int best_action_index = -1;
    for (int i = 0; i < (int)child_nodes_.size(); i++)
    {
        const auto &child_node = child_nodes_[i];
        double ucb1_value = child_node.ucb1(t);
        if (ucb1_value > best_value)
        {
            best_action_index = i;
            best_value = ucb1_value;
        }
    }
    return child_nodes_[best_action_index];
}

void Node::print_tree(const int depth) const
{
    using std::cout;
    using std::endl;

    for (int i = 0; i < (int)child_nodes_.size(); i++)
    {
        const auto &child_node = child_nodes_[i];
        for (int j = 0; j < depth; j++)
            cout << "__";
        cout << " " << i << "(" << child_node.n_ << ")" << endl;
        if (!child_node.child_nodes_.empty())
        {
            child_node.print_tree(depth + 1);
        }
    }
}
